File: examples3d/debug_dynamic_collider_add3.rs
use rapier_testbed3d::Testbed;
use rapier3d::prelude::*;

pub fn init_world(testbed: &mut Testbed) {
    /*
     * World
     */
    let mut bodies = RigidBodySet::new();
    let mut colliders = ColliderSet::new();
    let impulse_joints = ImpulseJointSet::new();
    let multibody_joints = MultibodyJointSet::new();

    /*
     * Ground.
     */
    let ground_size = 20.0;
    let ground_height = 0.1;

    let rigid_body = RigidBodyBuilder::fixed().translation(vector![0.0, -ground_height, 0.0]);
    let ground_handle = bodies.insert(rigid_body);
    let collider = ColliderBuilder::cuboid(ground_size, ground_height, 0.4)
        .friction(0.15)
        // .restitution(0.5)
        ;
    colliders.insert_with_parent(collider, ground_handle, &mut bodies);

    /*
     * Rolling ball
     */
    let ball_rad = 0.1;
    let rb = RigidBodyBuilder::dynamic()
        .translation(vector![0.0, 0.2, 0.0])
        .linvel(vector![10.0, 0.0, 0.0]);
    let ball_handle = bodies.insert(rb);
    let collider = ColliderBuilder::ball(ball_rad).density(100.0);
    colliders.insert_with_parent(collider, ball_handle, &mut bodies);

    let mut linvel = Vector::zeros();
    let mut angvel = Vector::zeros();
    let mut pos = Isometry::identity();
    let mut step = 0;
    let mut extra_colliders = Vec::new();
    let snapped_frame = 51;

    testbed.add_callback(move |mut graphics, physics, _, _| {
        step += 1;

        // Add a bigger ball collider
        let collider = ColliderBuilder::ball(ball_rad + 0.01 * (step as f32)).density(100.0);
        let new_ball_collider_handle =
            physics
                .colliders
                .insert_with_parent(collider, ball_handle, &mut physics.bodies);

        if let Some(graphics) = &mut graphics {
            graphics.add_collider(new_ball_collider_handle, &physics.colliders);
        }

        extra_colliders.push(new_ball_collider_handle);

        // Snap the ball velocity or restore it.
        let ball = physics.bodies.get_mut(ball_handle).unwrap();

        if step == snapped_frame {
            linvel = *ball.linvel();
            angvel = *ball.angvel();
            pos = *ball.position();
        }

        if step == 100 {
            ball.set_linvel(linvel, true);
            ball.set_angvel(angvel, true);
            ball.set_position(pos, true);
            step = snapped_frame;

            for handle in &extra_colliders {
                if let Some(graphics) = &mut graphics {
                    graphics.remove_collider(*handle, &physics.colliders);
                }

                physics
                    .colliders
                    .remove(*handle, &mut physics.islands, &mut physics.bodies, true);
            }

            extra_colliders.clear();
        }

        // Remove then re-add the ground collider.
        // let ground = physics.bodies.get_mut(ground_handle).unwrap();
        // ground.set_position(Isometry::translation(0.0, step as f32 * 0.001, 0.0), false);
        // let coll = physics
        //     .colliders
        //     .remove(ground_collider_handle, &mut physics.bodies, true)
        //     .unwrap();
        let coll = ColliderBuilder::cuboid(ground_size, ground_height + step as f32 * 0.01, 0.4)
            .friction(0.15);
        let new_ground_collider_handle =
            physics
                .colliders
                .insert_with_parent(coll, ground_handle, &mut physics.bodies);

        if let Some(graphics) = &mut graphics {
            graphics.add_collider(new_ground_collider_handle, &physics.colliders);
        }

        extra_colliders.push(new_ground_collider_handle);
    });

    /*
     * Set up the testbed.
     */
    testbed.set_world(bodies, colliders, impulse_joints, multibody_joints);
    testbed.look_at(point![10.0, 10.0, 10.0], Point::origin());
}

File: python/rapier-testbed/rapier_testbed/examples3/debug_dynamic_collider_add3.py
"""Port of ``examples3d/debug_dynamic_collider_add3.rs``.

A rolling ball whose collider keeps growing each frame while extra ground
colliders are also added. Snaps the ball's velocity at frame 51, restores
it at frame 100, and removes the extra colliders.
"""
from __future__ import annotations

import rapier3d as rp

from .._registry import register

CATEGORY = "Debug"
NAME = "Dyn. collider add"


def init_world(testbed) -> None:
    bodies = rp.RigidBodySet()
    colliders = rp.ColliderSet()
    impulse_joints = rp.ImpulseJointSet()
    multibody_joints = rp.MultibodyJointSet()

    ground_size = 20.0
    ground_height = 0.1
    ground_body = rp.RigidBody.fixed().translation((0.0, -ground_height, 0.0))
    ground_handle = bodies.insert(ground_body)
    colliders.insert_with_parent(
        rp.Collider.cuboid(ground_size, ground_height, 0.4).friction(0.15),
        ground_handle,
        bodies,
    )

    ball_rad = 0.1
    ball_body = (
        rp.RigidBody.dynamic()
        .translation((0.0, 0.2, 0.0))
        .linvel((10.0, 0.0, 0.0))
    )
    ball_handle = bodies.insert(ball_body)
    colliders.insert_with_parent(
        rp.Collider.ball(ball_rad).density(100.0), ball_handle, bodies
    )

    state = {
        "linvel": rp.Vec3(0.0, 0.0, 0.0),
        "angvel": rp.Vec3(0.0, 0.0, 0.0),
        "pos": rp.Isometry3.identity(),
        "step": 0,
        "extras": [],
    }
    snapped_frame = 51

    def _cb(tb) -> None:
        state["step"] += 1
        step = state["step"]

        # Add a bigger ball collider.
        new_handle = tb.colliders.insert_with_parent(
            rp.Collider.ball(ball_rad + 0.01 * step).density(100.0),
            ball_handle,
            tb.bodies,
        )
        state["extras"].append(new_handle)

        ball = tb.bodies.get(ball_handle)
        if ball is None:
            return
        if step == snapped_frame:
            state["linvel"] = ball.linvel
            state["angvel"] = ball.angvel
            state["pos"] = ball.position
        if step == 100:
            ball.linvel = state["linvel"]
            ball.angvel = state["angvel"]
            ball.position = state["pos"]
            state["step"] = snapped_frame
            for h in state["extras"]:
                tb.colliders.remove(h, tb._islands, tb.bodies, True)
            state["extras"].clear()

        # Add a fresh ground collider too.
        new_ground = tb.colliders.insert_with_parent(
            rp.Collider.cuboid(ground_size, ground_height + step * 0.01, 0.4).friction(0.15),
            ground_handle,
            tb.bodies,
        )
        state["extras"].append(new_ground)

    testbed.add_callback(_cb)

    testbed.set_world(bodies, colliders, impulse_joints, multibody_joints)
    testbed.look_at((10.0, 10.0, 10.0), (0.0, 0.0, 0.0))


register(CATEGORY, NAME, init_world, dim=3)


if __name__ == "__main__":
    from .._picker import run

    run(initial=f"{CATEGORY} / {NAME}")
